README.md
# The Link Crab

[![PyPI version](https://badge.fury.io/py/link-crab.svg)](https://badge.fury.io/py/link-crab)
![Run Pytest](https://github.com/klucsik/link-crab/workflows/Run%20Pytest/badge.svg)
[![Coverage Status](https://coveralls.io/repos/github/klucsik/link-crab/badge.svg?branch=master)](https://coveralls.io/github/klucsik/link-crab?branch=master)

A simple CLI tool which can crawl through your website and catch broken links, and can check user permissions to specific pages on your website.

## Workmode - Link gathering:
In this mode, you provide a starting url, and the Link Crab will crawl through the starting page, and all the page which is accessible thorugh links from that page and is in the same domain as the starting apge.
The program export the gathered links in a txt file, then exercise them, gathering response time and status code, and exporting these in a csv file.

 ## Workmode - Link access permission checking:
In this mode you provide a csv file with links to check, and wether those links should be accessible. The Link Crab will check every link in the list, determines if its accessible or not, and then assert the expected accessibilty to the actual accessibility. 
A link is considered accessible if the http response for a get request on the link has a status code under 400, and after all redirects, the url is equals of the starting url. 
(Most of the websites either give you a 404 or redirect to the sign-in page.)
*Maybe following the redirects is unnecessary, but I considered it safer*

## Session management:
In both workmode, you can provide login informations. The Link Crab opens up a Chrome browser with Selenium webdriver, and goes to the provided login page. On the login page, it will find the email and password fields, based on the html ids you provided, and fills it with your credentials. Then it will click on the submit button, thus logs in to the page. Then the Link Crab aquires the cookies, and closes the browser. It will use this auth cookies through the testing.

If you are in Permission checking mode, and want to check the logout page, be sure you will not need the session in the next checked pages after the logout.

## Generated reports:

The Link Crab will make the following reports:

**[domain_name]_links.txt:**
This report is generated in the link gathering mode, through the gathering phase. It will be updated as the gatherer crawls through the pages.

Example from the mock app:

    http://127.0.0.1:5000/user/sign-in
    http://127.0.0.1:5000/user/forgot-password
    http://127.0.0.1:5000/members
    http://127.0.0.1:5000/user/edit_user_profile
    http://127.0.0.1:5000/missing
    http://127.0.0.1:5000/user/change-password
    http://netdna.bootstrapcdn.com/bootstrap/3.1.1/css/bootstrap.min.css
    http://127.0.0.1:5000/user/missing
    http://127.0.0.1:5000/user/register
    http://127.0.0.1:5000/user/sign-out
    http://127.0.0.1:5000/
    http://127.0.0.1:5000/missing_member_only
    http://127.0.0.1:5000/admin

**[domain_name]_[datetime]_exercised_links.csv:**
This report is generated at the end of the exercising phase in the link gathering mode.

Example from the mock app:

| url                                                                  | status_code | resp_url(after_redirects)                                            | response_time(ms) | accessible? |
|----------------------------------------------------------------------|-------------|----------------------------------------------------------------------|-------------------|-------------|
| http://127.0.0.1:5000/user/sign-in                                   | 200         | http://127.0.0.1:5000/                                               | 10                | False       |
| http://127.0.0.1:5000/user/forgot-password                           | 200         | http://127.0.0.1:5000/user/forgot-password                           | 6                 | True        |
| http://127.0.0.1:5000/members                                        | 200         | http://127.0.0.1:5000/members                                        | 10                | True        |
| http://127.0.0.1:5000/user/edit_user_profile                         | 200         | http://127.0.0.1:5000/user/edit_user_profile                         | 6                 | True        |
| http://127.0.0.1:5000/missing                                        | 404         | http://127.0.0.1:5000/missing                                        | 3                 | False       |
| http://127.0.0.1:5000/user/change-password                           | 200         | http://127.0.0.1:5000/user/change-password                           | 13                | True        |
| http://netdna.bootstrapcdn.com/bootstrap/3.1.1/css/bootstrap.min.css | 200         | http://netdna.bootstrapcdn.com/bootstrap/3.1.1/css/bootstrap.min.css | 94                | True        |
| http://127.0.0.1:5000/user/missing                                   | 404         | http://127.0.0.1:5000/user/missing                                   | 2                 | False       |
| http://127.0.0.1:5000/user/register                                  | 200         | http://127.0.0.1:5000/user/register                                  | 7                 | True        |
| http://127.0.0.1:5000/user/sign-out                                  | 200         | http://127.0.0.1:5000/                                               | 8                 | False       |
| http://127.0.0.1:5000/                                               | 200         | http://127.0.0.1:5000/                                               | 8                 | True        |
| http://127.0.0.1:5000/missing_member_only                            | 404         | http://127.0.0.1:5000/missing_member_only                            | 3                 | False       |
| http://127.0.0.1:5000/admin                                          | 200         | http://127.0.0.1:5000/user/sign-in?next=/admin                       | 9                 | False       |

**[domain_name]_[datetime]_[user_email]_permission_check_result.csv:**
This report is generated at the end of the permission checking mode.

Example from the mock app:

| url                                          | status_code | resp_url(after_redirects)                                     | accessible? | should_be_accessible? | assert_accessibility |
|----------------------------------------------|-------------|---------------------------------------------------------------|-------------|-----------------------|----------------------|
| http://127.0.0.1:5000/                       | 200         | http://127.0.0.1:5000/                                        | True        | True                  | PASSED               |
| http://127.0.0.1:5000/user/register          | 200         | http://127.0.0.1:5000/user/register                           | True        | True                  | PASSED               |
| http://127.0.0.1:5000/members                | 200         | http://127.0.0.1:5000/members                                 | True        | True                  | PASSED               |
| http://127.0.0.1:5000/user/forgot-password   | 200         | http://127.0.0.1:5000/user/forgot-password                    | True        | True                  | PASSED               |
| http://127.0.0.1:5000/user/edit_user_profile | 200         | http://127.0.0.1:5000/user/edit_user_profile                  | True        | True                  | PASSED               |
| http://127.0.0.1:5000/admin                  | 200         | http://127.0.0.1:5000/                                        | False       | False                 | PASSED               |
| http://127.0.0.1:5000/user/sign-in           | 200         | http://127.0.0.1:5000/                                        | False       | True                  | FAILED               |
| http://127.0.0.1:5000/user/sign-out          | 200         | http://127.0.0.1:5000/                                        | False       | True                  | FAILED               |
| http://127.0.0.1:5000/user/change-password   | 200         | http://127.0.0.1:5000/user/sign-in?next=/user/change-password | False       | True                  | FAILED               |


All reports are saved in the reports folder under a folder named by the domain name. For example, when you test  `example.com`, the reports will be in `reports/example.com/` relative to where you called the command.

The configuration is done through a yaml config files.

## Installation

Install with `pip install link-crab`

Dependencies: [chromedriver](https://chromedriver.chromium.org/downloads) for logging in to the tested site.

## Usage:
Simply use the command `python -m link_crab path/to/your/config.yaml` in the PYthon envrionment which has the link-crab installed. All the configuration is done in the config file, which is expanded bellow.
If you want to use the sample flask mock app for testing, provide the `-t` flag.

For additional help run:  `python -m link-crab -h`

A good usage pattern would be to run the Link Crab first in link gathering mode, and from the generated links.txt select the links for the permission checking mode.

### Configuration:
**starting_url**

    starting_url: http://127.0.0.1:5000

Gather the reachable links in the starting_url's page and all of its subpages.
After collecting all the links, the link exerciser load every in-domain url with a GET request, and measures 
status code, response time, response url after all redirects, and accessibility based on status code and response url

**path_to_link_perms**

    path_to_link_perms: testapp_member_access.csv

Test accessibility of provided links. The csv should have a link and a should-access column. 
asserts the link accessibility equals to provided should-access.
A link is accessible if the response status code<400, and after redirets the respone url equals the starting url
(some framework give a 404 for unaccessible pages or redriects to sign_in page)

 Sample link_perms csv:

| link                                         | should-access |
|----------------------------------------------|---------------|
| http://127.0.0.1:5000/                       | TRUE          |
| http://127.0.0.1:5000/user/register          | TRUE          |
| http://127.0.0.1:5000/members                | TRUE          |
| http://127.0.0.1:5000/admin                  | FALSE         |


**User**

    user:
        email: member@example.com
        email_locator_id: email
        login_url: http://127.0.0.1:5000/user/sign-in
        password: Password1
        password_locator_id: password
    
Login with the help of selenium webdriver (chromedriver). You need to provide the url of the login form, 
   and the id's of the email (or username) and password fields.

CHANGELOG.md
# release v0.2.1

  * DOCUMENTATION: Fixed table formatting in Generated Reports section

 # release v0.2.0

 * TECHNICAL FEATURE: common test fixtures moved out to helper files, session namecollision fixed.
 * FEATURE: Better logging
 * FEATURE: Generate summary at the end of the reports
 * DOCUMENTATION: Fixed table formatting in Generated Reports section

# release v0.1.1

* FEATURE: exercise report name now contains the username
* BUGFIX: exercise report header was unaligned
* TECHNICAL FEATURE: Add GitHub Actions CI for automated testing and coverage report
* TECHNICAL FEATURE: Added Pipy badge to readme
* BUGFIX: Fixed outcome for the exception path of the excersizeurl function *@tothmate92*

# release v0.1.0

link_crab/__init__.py
import os
import sys
sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
from tests.test_app import generate_mock_app
import time
import link_crab.session_manager
import link_crab.reporting
import link_crab.exercise_url
import link_crab.gather_links
import link_crab.csv_reader

import logging

from urllib.request import urlparse
import yaml
from datetime import datetime
import colorama

# Colorama init https://pypi.org/project/colorama/
colorama.init()
GREEN = colorama.Fore.GREEN
RED = colorama.Fore.RED
YELLOW = colorama.Fore.YELLOW
GRAY = colorama.Fore.LIGHTBLACK_EX
CYAN = colorama.Fore.CYAN
RESET = colorama.Fore.RESET

# CLI interface:
def main():
    import argparse
    parser = argparse.ArgumentParser(
        description="Link-Crab: A link crawler and permission testing tool, written in Python",
        usage=f"""
        Provide a path for your config yaml file. 
        Usable {CYAN}config keys{RESET}:
            {CYAN}starting_url: http://127.0.0.1:5000{RESET}
                Gather the reachable links in the starting_url's page and all of its subpages.
                After collecting all the links, the link exerciser load every in-domain url with a GET request, and measures 
                status code, response time, response url after all redirects, and accessibility based on status code and response url
            
            {CYAN}path_to_link_perms: testapp_member_access.csv{RESET}
                Test accessibility of provided links. The csv should have a link and a should-access column. 
                asserts the link accessibility equals to provided should-access.
                A link is accessible if the response status code<400, and after redirets the respone url equals the starting url
                (some framework give a 404 for unaccessible pages or redriects to sign_in page)

            {CYAN}user:
                email: member@example.com
                email_locator_id: email
                login_url: http://127.0.0.1:5000/user/sign-in
                password: Password1
                password_locator_id: password{RESET}
                    Login with the help of selenium webdriver (chromedriver). You need to provide the url of the login form, 
                    and the id's of the email (or username) and password fields.
        """
        )
    parser.add_argument("config_yaml_path", metavar='config_yaml_path', type=str,
                        help="Path to the config yaml file.")
    parser.add_argument("-t" ,"--test", action='store_true', help="wind up the default test application for testing")
    parser.add_argument("-v","--verbose", action='store_true', help="sets verbosity level for the cli output.")

    args = parser.parse_args()
    config_yaml_path = args.config_yaml_path

    # setup:
    starting_time = datetime.now()
    config = {}
    with open(config_yaml_path) as f:
        config = yaml.load(f, Loader=yaml.FullLoader)

    starting_url = None
    try:
        starting_url = config['starting_url']
    except:
        pass

    path_to_link_perms = None
    try:
        path_to_link_perms = config['path_to_link_perms']
    except:
        pass

    checked_domain=None
    if starting_url:
        checked_domain = urlparse(starting_url).netloc
    if path_to_link_perms:
        link_perms = csv_reader.read_link_perms(path_to_link_perms)
        checked_domain = urlparse(link_perms[0][0]).netloc
    
    #logger.filehandler
    if not os.path.exists(f'reports/{checked_domain}'):
        os.makedirs(f'reports/{checked_domain}')  

    logger = logging.getLogger(__name__)
    logger.setLevel(logging.DEBUG)
    fh = logging.FileHandler(f'reports/{checked_domain}/{checked_domain}_{starting_time}.log')
    fh.setLevel(logging.DEBUG)

    #lofgger.consolehandler
    ch = logging.StreamHandler()
    if args.verbose:
        ch.setLevel(logging.DEBUG)
        print("being Verbose!")
    else:
        ch.setLevel(logging.INFO)

    formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
    fh.setFormatter(formatter)
    ch.setFormatter(formatter)

    logger.addHandler(fh)
    logger.addHandler(ch)

    logger.info('----------------setup---------------------')
    logger.info(f"checked_domain: {checked_domain}")
   
    if args.test:
        generate_mock_app()

    user = None
    try:
        user = config['user']
    except:
        pass
    session = session_manager.make_session(user)

    links = set()

    links.add(starting_url)

    exit_value = 0
    link_error_count = 0
    perm_error_count = 0 

        
    # gathering:
    if starting_url:
        logger.info('----------------gather links---------------------')
        
        links.add(starting_url)
        links = gather_links.gather_links(session, links, checked_domain)

    # exercising:
        logger.info('----------------Excercise links---------------------')
        link_db = []
        # remake the session, because crawling through the log-out link logs us out :D
        session = session_manager.make_session(user)
        for link in links:
            exercise_outcome = exercise_url.exercise_url(session, link)
            link_db.append(exercise_outcome)
            if exercise_outcome[1] > 400:
                exit_value = 1
                link_error_count = link_error_count + 1

        reporting.save_linkdb_to_csv(link_db, checked_domain,starting_time)

        logger.debug('----------------REPORT: Excercised links---------------------')
        for link in link_db:
            logger.debug(f"[*] {link[0]} - {link[1]} - {link[4]} - {link[2]} ms - accessible: {link[3]}")

    # permission checking:

    
    if path_to_link_perms:
        logger.info('----------------permission checking---------------------')
        link_perm_db = []
        link_perms = csv_reader.read_link_perms(path_to_link_perms) # TODO: remove as duplicate
        checked_domain = urlparse(link_perms[0][0]).netloc # TODO: remove as duplicate
        session = session_manager.make_session(user)
        for link_perm in link_perms:
            outcome = (exercise_url.exercise_url(session, link_perm[0]))
            assertion_outcome = 'FAILED'
            if outcome[3] == link_perm[1]:
                assertion_outcome = 'PASSED'
            outcome = outcome + [link_perm[1]] + [assertion_outcome]
            link_perm_db.append(outcome)

        reporting.save_permdb_to_csv(link_perm_db, checked_domain, user['email'],starting_time)

        logger.debug('----------------REPORT: Checked permissions---------------------')
        for link in link_perm_db:
            if link[6] != 'PASSED':
                exit_value = 1
                perm_error_count = perm_error_count + 1
            logger.debug(f"[*] {link[0]} - {link[1]} - {link[4]} - {link[2]} ms - accessible: {link[3]} - should-be?: {link[5]} - assert: {link[6]}")

    # Exit:
    if starting_url:
        print(f"Gathered {len(link_db)} links, {link_error_count} of them broken.")
    if path_to_link_perms:
        print(f"Checked {len(link_perm_db)} links for permission, {perm_error_count} of them is not met the expectations.")
    print(f"exit value:{exit_value}")
    sys.exit(exit_value)



